src/ray/core_worker/task_execution.cc

```diff
@@ -7,65 +7,63 @@ namespace ray {
-      std::unique_ptr<CoreWorkerRayletTaskReceiver>(new CoreWorkerRayletTaskReceiver(
-          raylet_client, main_service_, worker_server_)));
+      std::unique_ptr<CoreWorkerRayletTaskReceiver>(
+          new CoreWorkerRayletTaskReceiver(main_service_, worker_server_, func)));
 
   // Start RPC server after all the task receivers are properly initialized.
   worker_server_.Run();
 }
 
-Status CoreWorkerTaskExecutionInterface::Run(const TaskExecutor &executor) {
-  while (true) {
-    std::vector<TaskSpec> tasks;
-    auto status =
-        task_receivers_[static_cast<int>(TaskTransportType::RAYLET)]->GetTasks(&tasks);
-    if (!status.ok()) {
-      RAY_LOG(ERROR) << "Getting task failed with error: "
-                     << ray::Status::IOError(status.message());
-      return status;
-    }
+Status CoreWorkerTaskExecutionInterface::ExecuteTask(
+    const raylet::TaskSpecification &spec) {
+  worker_context_.SetCurrentTask(spec);
 
-    for (const auto &task : tasks) {
-      const auto &spec = task.GetTaskSpecification();
-      worker_context_.SetCurrentTask(spec);
-
-      RayFunction func{spec.GetLanguage(), spec.FunctionDescriptor()};
-
-      std::vector<std::shared_ptr<RayObject>> args;
-      RAY_CHECK_OK(BuildArgsForExecutor(spec, &args));
-
-      TaskType task_type;
-      if (spec.IsActorCreationTask()) {
-        task_type = TaskType::ACTOR_CREATION_TASK;
-      } else if (spec.IsActorTask()) {
-        task_type = TaskType::ACTOR_TASK;
-      } else {
-        task_type = TaskType::NORMAL_TASK;
-      }
-
-      TaskInfo task_info{spec.TaskId(), spec.JobId(), task_type};
-
-      auto num_returns = spec.NumReturns();
-      if (spec.IsActorCreationTask() || spec.IsActorTask()) {
-        RAY_CHECK(num_returns > 0);
-        // Decrease to account for the dummy object id.
-        num_returns--;
-      }
-
-      status = executor(func, args, task_info, num_returns);
-      // TODO(zhijunfu):
-      // 1. Check and handle failure.
-      // 2. Save or load checkpoint.
-    }
+  RayFunction func{spec.GetLanguage(), spec.FunctionDescriptor()};
+
+  std::vector<std::shared_ptr<RayObject>> args;
+  RAY_CHECK_OK(BuildArgsForExecutor(spec, &args));
+
+  TaskType task_type;
+  if (spec.IsActorCreationTask()) {
+    task_type = TaskType::ACTOR_CREATION_TASK;
+  } else if (spec.IsActorTask()) {
+    task_type = TaskType::ACTOR_TASK;
+  } else {
+    task_type = TaskType::NORMAL_TASK;
+  }
+
+  TaskInfo task_info{spec.TaskId(), spec.JobId(), task_type};
+
+  auto num_returns = spec.NumReturns();
+  if (spec.IsActorCreationTask() || spec.IsActorTask()) {
+    RAY_CHECK(num_returns > 0);
+    // Decrease to account for the dummy object id.
+    num_returns--;
   }
 
+  auto status = execution_callback_(func, args, task_info, num_returns);
+  // TODO(zhijunfu):
+  // 1. Check and handle failure.
+  // 2. Save or load checkpoint.
+  return status;
+}
+
+Status CoreWorkerTaskExecutionInterface::Run() {
```

Comment:
Is the return status still needed?
And how can we propagate errors to Java now?

Reply:
the status is not needed and can be removed. 
we're not propagating errors to java here, previously we just check the return value of the java/python callback, and return directly if there's an error.